docs/specs/oidc-native-sso.md

```diff
@@ -0,0 +1,155 @@
+- A OfflineGrant used to has only one set of client-specific information, like `client_id` and `refresh_token`. Now, a OfflineGrant can have multiple refresh tokens.
+- Since there is only one OfflineGrant, the apps sharing user authentication with Native SSO does not share refresh token. But the underlying session is shared. Thus signing out in one app will sign out all apps. This is by design. See [App2app](./app2app.md) if you want the apps have independent sessions.
+- Native SSO is done through the Token Endpoint, thus it requires no user interaction.
+- Existing OfflineGrant cannot perform Native SSO. It is because `device_sso` is not in the `scope`. Sign in again to obtain a Native SSO OfflineGrant to perform Native SSO.
+
+## Changes on Client
+
+- Add `x_device_sso_enabled: boolean`.
+- `scope=device_sso` is allowed if `x_device_sso_enabled=true`.
+
+> Do we need to add `x_device_sso_key: string` to designate which group of clients can perform Native SSO?
+> Only clients with `x_device_sso_enabled=true` AND the same value of `x_device_sso_key` can perform Native SSO with each other.
+> This seems very advanced to me.
+
+## Changes on OfflineGrant
+
+- The following fields become client-specific
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
+  - `Scopes`
+  - `TokenHash`
+- Add a new field `DeviceSSORefreshTokens`. It has the above client-specific fields.
+  - `ClientID`
+  - `AuthorizationID`
+  - `IdentityID`
+  - `Scopes`
+  - `TokenHash`
+- Add a new field `DeviceSecretHash`. It is the hex of SHA256 of `device_secret`.
+- If `DeviceSSORefreshTokens` is non-empty, then the offline grant is a Native SSO offline grant.
+- Otherwise, the offline grant is an ordinary offline grant.
+
+## Changes on Authorization Endpoint
+
+- Allow `scope=device_sso` if the client has `x_device_sso_enabled=true`.
+
+## Changes on Token Endpoint
+
+### `grant_type=authorization_code`
+
+- If `authorization_code.scope=device_sso` and `device_secret` is present and it is valid, a new `refresh_token` is added to Native SSO offline grant.
+- If `authorization_code.scope=device_sso` and `device_secret` is absent or it is invalid, a new Native SSO offline grant is created.
+
+### `grant_type=refresh_token`
+
+- If `refresh_token.scope=device_sso` and `device_secret` is present and it is valid, nothing to do.
+- If `refresh_token.scope=device_sso` and `device_secret` is absent or it is invalid, a new `device_secret` is generated. `DeviceSecretHash` is updated.
+
+### `grant_type=urn:authgear:params:oauth:grant-type:biometric-request`
+
+- Support a new parameter `scope`.
+- Allow `scope=device_sso` if the client has `x_device_sso_enabled=true`.
+- If `scope=device_sso` and `device_secret` is present and it is valid, a new `refresh_token` is added to Native SSO offline grant.
+- If `scope=device_sso` and `device_secret` is absent or it is invalid, a new Native SSO offline grant is created.
+
+### `grant_type=urn:authgear:params:oauth:grant-type:app2app`
```

Comment:
I suggest we do not support native sso with app2app. So we don't allow specifying scope in this grant_type.

Or else, we need to ensure the provided scope is a subset of the scope of the original offline grant.

Reply:
> we need to ensure the provided scope is a subset of the scope of the original offline grant

Let's do this.

Reply:
I added a commit to clarify this.